Design doc for upsert order by timestamp (#18722)
MaterializeInc/materialize: doc/developer/design/20230411_envelope_upsert_order_by_timestamp.md

- Feature name: Envelope upsert with order by timestamp (limited scope)
- Associated: https://github.com/MaterializeInc/materialize/issues/16512

# Summary

- Support user specified order by in Kafka upsert sources in a limited capacity. If users want to specify custom ordering which is semantically meaningful to them, there's currently no easy way to do that. They could potentially use `ENVELOPE NONE` along with a `DISTINCT ON` view to get similar results, but it's cumbersome and lacks the automatic compaction on keys with normal `ENVELOPE UPSERT`.
- Currently, will limit the scope to allow only `ORDER BY` the `TIMESTAMP` metadata field. Depending upon user feedback this can later be expanded to support expressions and extractions from other columns like `HEADERS`.

# Scope

The scope will be limited to only allow `ORDER BY TIMESTAMP` (ascending order).

> - For kafka upsert style topics, value = NULL is taken to mean that it’s a delete tombstone and that the corresponding key should be removed. If we want to keep the same behavior it would make sense to limit the `ORDER BY` columns to only the metadata columns.

Reviewer: What's an upsert-style topic, and how is it distinct from a key-compacted topic?

(We've declared the latter out of scope, which I think is wise. But Kafka itself doesn't ascribe any semantics to null values outside the key-compacted case, and it's not clear to me that it's ever actually a thing people do.)

Author: By "upsert style" I meant topics having the semantics so that it could be compacted on some value for a key. 

I would imagine key-compacted to be a kafka specific definition where it can be only compacted on the latest offset value. I guess from that POV, the "upsert style" is more of a looser definition that it can have a different field to compact things by (in this case, timestamp).

Regarding ascribing meaning to null values, it's easier to borrow the definition of tombstones since it already works in a similar fashion for current upsert?

Reviewer: Yeah, I don't think I disagree with any of that! But as-is it's a bit of a circular definition.

It seems possible that we could avoid introducing the concept of an upsert-style topic at all:

```suggestion
- In a key-compacted Kafka copic, value = NULL is taken to mean that it’s a delete tombstone and that the corresponding key should be removed. If we want to borrow that tombstone semantic for our more flexible upsert source, it would make sense to limit the `ORDER BY` columns to only the metadata columns.
```

Otherwise we should define it! (But it seems like we're inventing something instead of adopting something that we found "in the wild", so I'd be inclined to skip it.)

Author: Ah I see what you mean. This makes more sense 👍

> - In a key-compacted Kafka copic, value = NULL is taken to mean that it’s a delete tombstone and that the corresponding key should be removed. If we want to borrow that tombstone semantic for our more flexible upsert source, it would make sense to limit the `ORDER BY` columns to only the metadata columns.

Reviewer 2: ```suggestion
- In a key-compacted Kafka topic, value = NULL is taken to mean that it’s a delete tombstone and that the corresponding key should be removed. If we want to borrow that tombstone semantic for our more flexible upsert source, it would make sense to limit the `ORDER BY` columns to only the metadata columns.
```

Author: thanks 🙈

    - `PARTITION`, `KEY`, `TOPIC`, would not make for meaningful orderings:
        - The `PARTITION` of a record is typically derived from its key, and so is the same for every record with a given key.
        - `KEY` is, by definition, the same for every record with a given key.
        - `TOPIC` is the same for every record in the topic.
    - `HEADERS` by itself would not also make a good order by clause (it’s a json blob), but later we can probably allow expressions so that values could be extracted from it.
- Limiting to only ascending order because no users have expressed a need for descending.

# Explanation

## Proposal

Make it possible with unsafe-mode to provide custom ORDER BY clause which will override the default ordering by offset. The given ORDER BY identifiers should refer to only the included TIMESTAMP and OFFSET metadata columns. It will be required that the user always mention OFFSET as one of the columns for the ORDER BY clause.

The option will be part of the `ENVELOPE UPSERT` clause with the following grammar:

> `ENVELOPE UPSERT [(ORDER BY (<expr>) [ASC])]`

Reviewer 3: With the change to support multiple columns, this should be updated to:

```
ENVELOPE UPSERT [(ORDER BY (<expr> [ASC], ...))]
```

To match the way this works for `SELECT`, the `ASC` should be optionally attached to _each_ expression.

Author: Ah I see what you mean. I saw `ORDER BY (col1, col2) ASC` is legal as well in select statements, so had kept that (I guess it's the same thing as you mentioned where the entire tuple is the expression). I will change it to ordering per expression.

The `ASC` modifier is optional noise for specifying ascending ordering, for symmetry with the `ORDER BY` clause in `SELECT` statements.

Examples of valid syntax and semantics:

> - `CREATE SOURCE ... INCLUDE TIMESTAMP ENVELOPE UPSERT ( ORDER BY ( TIMESTAMP, OFFSET ) ASC )`

Reviewer 3: ```
CREATE SOURCE ... INCLUDE TIMESTAMP ENVELOPE UPSERT ( ORDER BY ( TIMESTAMP  ASC, OFFSET ASC) )
```

Author: I will update the design doc in follow up implementation PR https://github.com/MaterializeInc/materialize/pull/18567

Author: Btw, I was trying out select statements in materialize. Parentheses around expressions with ordering is erroring.

```
 select * from texttext order by (key ASC, text DESC);
ERROR:  Expected right parenthesis, found ASC
LINE 1: select * from texttext order by (key ASC, text DESC);
```

The following works though
```
select * from texttext order by key ASC, text;

select * from texttext order by (key, text) ASC;
```

Author: Should we lose the parentheses around the multiple order by statements with order then? Or go back to limiting it to one tuple expression with optional ASC?

Reviewer 3: Right! The syntax for the `ORDER BY` clause in `SELECT` is:

```
select_order_by ::= ORDER BY <order_by_col> [, <order_by_col>]*
order_by_col := <expr> [ASC | DESC]
expr := <ident> | ( <expr> ) | ...
```

Whereas the syntax for the `ORDER BY` option in `CREATE SOURCE` will be:

```
create_source_option_order_by ::= ORDER BY (<order_by_col> [, <order_by_col>]*)
order_by_col := <expr> [ASC | DESC]
expr := <ident> | ( <expr> ) | ...
```

So the syntaxes have to _slightly_ deviate, because in `CREATE SOURCE` we need parens around the order by columns in order to disambiguate between a new `<order_by_col>` and a new `CREATE SOURCE` option.

The equivalent of

```
select * from texttext order by (key, text) ASC;
```

would be:

```
CREATE SOURCE ... ENVELOPE UPSERT (ORDER BY ((key, text) ASC))
```

Author: Makes sense. Thanks for explaining that!

- Feature name: Envelope upsert with order by timestamp (limited scope)
- Associated: https://github.com/MaterializeInc/materialize/issues/16512

# Summary

- Support user specified order by in Kafka upsert sources in a limited capacity. If users want to specify custom ordering which is semantically meaningful to them, there's currently no easy way to do that. They could potentially use `ENVELOPE NONE` along with `DISTINCT ON` to get similar results, but it's cumbersome and lacks the automatic compaction on keys with normal `ENVELOPE UPSERT`.
- Currently, will limit the scope to allow only `ORDER BY` the `TIMESTAMP` metadata field behind an unsafe flag. Depending upon user feedback this can later be expanded to support expressions and extractions from other columns like `HEADERS`.

# Scope

> The scope will be limited to only allow `ORDER BY TIMESTAMP` (default ordering `ASC`).

Reviewer 4: ```suggestion
The scope will be limited to only allow `ORDER BY TIMESTAMP` (ascending order).
```

- For kafka upsert style topics, value = NULL is taken to mean that it’s a delete tombstone and that the corresponding key should be removed. If we want to keep the same behavior it would make sense to limit the `ORDER BY` columns to only the metadata columns.
- Limiting it to only `TIMESTAMP` because
    - `OFFSET` is already default. We can make it explicit though if we want
    - `PARTITION`, `KEY`, `TOPIC`, would not make a good order by clause
    - `HEADERS` by itself would not also make a good order by clause (it’s a json blob), but later we can probably allow expressions so that values could be extracted from it.

> - Limiting to only `ASC` as default, because for compaction it makes sense to take semantically the _latest_ value of something, which implies an `ASC` order by

Reviewer 4: Are you talking about kafka compaction here? I think you can simplify it to this:

```suggestion
- Limiting to only ascending order because no users have expressed a need for descending.
```

Author: I was referring to the `upsert` itself (since we are sort of doing the compaction on the key space), but I think this update reads better.

Here's an example of the upsert behavior where the source consists of a key, value and the timestamp from metadata and is ordered by the timestamp.

```
 Previously persisted data        Incoming kafka stream decoded
+------------------------+         +-------------------------+
| Ok (key1, value1, 100) |         |  Ok (key1, value3, 200) |
| Ok (key2, value2, 201) |         |  Err(key3, some_error2) |
| Err(key3, some_error1) |         |  Ok (key2, value4, 200) |
+------------------+-----+         +-----+-------------------+
                   |                     |
                   |                     |
                   |                     |
                   |                     |
                   v    Upsert Output    v
                  +------------------------+
                  | Ok (key1, value3, 200) |
                  | Ok (key2, value2, 201) |
                  | Err(key3, some_error2) |
                  +------------------------+

> ```

Reviewer 4: great diagram!

Note: As shown in the example above, the errors are still implicitly ordered by offset as we do not persist any extra metadata for them separately. A later error with the same key will always overwrite a previous one.

# Rollout

> This feature will be rolled out behind the unsafe flag.

Reviewer 4: ```suggestion
This feature will be rolled out behind the unsafe flag, then upgraded to only being available behind a LaunchDarkly flag.
```

    - How is their data structured and what columns will they want to use for the custom order by?
    - What kind of compaction are they using?

# Drawbacks

Users will need to be careful of what kind of compaction they have for the kafka topic. The upsert will do the correct thing based on the data it sees.

If old data has been compacted away either via time based or key based compaction, and we start ingesting the source after that, the result of upsert might not be accurate.

If users actually are using this custom order by, it’s likely they are **not** using key based compaction with Kafka because semantically that would be upsert-ing on offset.

# Conclusion and alternatives

An alternative could be waiting for `TRANSFORM USING` which has a much broader scope and could cover this particular scenario. In the meantime, having a smaller scoped custom order by could be useful for interested users.

# Unresolved questions
- In case the order by value of two entries are same, should it chose one of them deterministically? For eg, say we already have (key1, value1, ts=100) and have incoming (key1, value2, ts=100). Should we override `value1` with `value2` or not?

# Future work

> Depends upon user feedback to see how this should be expanded.

Reviewer 4: I think you can just mention that adding support for user-provided columns is a future extension, but dealing with tombstones is an unresolved problem!

Author: Updated!

- Associated: https://github.com/MaterializeInc/materialize/issues/16512

# Summary

- Support user specified order by in Kafka upsert sources in a limited capacity. If users want to specify custom ordering which is semantically meaningful to them, there's currently no easy way to do that. They could potentially use `ENVELOPE NONE` along with a `DISTINCT ON` view to get similar results, but it's cumbersome and lacks the automatic compaction on keys with normal `ENVELOPE UPSERT`.
- Currently, will limit the scope to allow only `ORDER BY` the `TIMESTAMP` metadata field. Depending upon user feedback this can later be expanded to support expressions and extractions from other columns like `HEADERS`.

# Scope

The scope will be limited to only allow `ORDER BY TIMESTAMP` (ascending order).

- For kafka upsert style topics, value = NULL is taken to mean that it’s a delete tombstone and that the corresponding key should be removed. If we want to keep the same behavior it would make sense to limit the `ORDER BY` columns to only the metadata columns.
- Limiting it to only `TIMESTAMP` because
    - `OFFSET` is already default. We can make it explicit though if we want.

>     - `PARTITION`, `KEY`, `TOPIC`, would not make a good order by clause.

Reviewer 3: A bit more of an explanation about why these fields are not good order by clauses would be helpful!

```suggestion
    - `PARTITION`, `KEY`, `TOPIC`, would not make for meaningful orderings:
        - The `PARTITION` of a record is typically derived from its key, and so is the same for every record with a given key.
        - `KEY` is, by definition, the same for every record with a given key.
        - `TOPIC` is the same for every record in the topic.
```

    - `HEADERS` by itself would not also make a good order by clause (it’s a json blob), but later we can probably allow expressions so that values could be extracted from it.
- Limiting to only ascending order because no users have expressed a need for descending.

# Explanation

## Proposal

Make it possible with unsafe-mode to provide custom ORDER BY clause which will override the default ordering by offset. The given ORDER BY identifier should refer to the included TIMESTAMP metadata column.

> 

Reviewer 3: ```suggestion

The option will be part of the `ENVELOPE UPSERT` clause with the following grammar:

ENVELOPE UPSERT [(ORDER BY <expr> [ASC])]

The `ASC` modifier is optional noise for specifying ascending ordering, for symmetry with the `ORDER BY` clause in `SELECT` statements.
```

Reviewer 3: cc @MaterializeInc/surfaces for review

Reviewer 3: It just occurred to me that we should wrap <expr> in parens, to support future multi-column ordering keys, as in

```
ENVELOPE UPSERT (ORDER BY (c1, c2))
```

It'll look a little silly as

```
ENVELOPE UPSERT (ORDER BY (c1))
```

but it'll ultimately be more consistent.

(Without the inner parens, something like `ENVELOPE UPSERT (ORDER BY c1, OTHER OPTION blah)` becomes ambiguous.)

> Examples of valid syntax:

Reviewer 3: ```suggestion
Examples of valid syntax and semantics:
```

- `CREATE SOURCE ... INCLUDE TIMESTAMP ENVELOPE UPSERT ( ORDER BY TIMESTAMP )`
- `CREATE SOURCE ... INCLUDE TIMESTAMP AS ts ENVELOPE UPSERT ( ORDER BY ts )`

> Examples of invalid syntax:

Reviewer 3: ```suggestion
Examples that are syntatically valid but semantically invalid:
```

(We maintain a strict separation between syntax errors—those which the parser rejects—and semantic errors—those which the SQL planner rejects.)

Here's an example of the upsert behavior where the source consists of a key, value and the timestamp from metadata and is ordered by the timestamp.

```
 Previously persisted data        Incoming kafka stream decoded
+------------------------+         +-------------------------+
| Ok (key1, value1, 100) |         |  Ok (key1, value3, 200) |
| Ok (key2, value2, 201) |         |  Err(key3, some_error2) |
| Err(key3, some_error1) |         |  Ok (key2, value4, 200) |
+------------------+-----+         +-----+-------------------+
                   |                     |
                   |                     |
                   |                     |
                   |                     |
                   v    Upsert Output    v
                  +------------------------+
                  | Ok (key1, value3, 200) |
                  | Ok (key2, value2, 201) |
                  | Err(key3, some_error2) |
                  +------------------------+
```

> Note: As shown in the example above, the errors are still implicitly ordered by offset as we do not persist any extra metadata for them separately. A later error with the same key will always overwrite a previous one.

Reviewer 3: This is surprising, IMO! Perhaps we _should_ be attaching the timestamp column to the errors, too? I don't actually understand in what case we can end up with `Err((key, error_text))`—if Avro decoding of the value fails, but not the key?

Reviewer 5: > if Avro decoding of the value fails, but not the key?
yes!

For some context, a somewhat recent change that made errors retractable: https://github.com/MaterializeInc/materialize/pull/14209. Which also made how errors work very complex.

Errors in the upsert source are a tricky subject. In general, the current "design" (evolved through quite some iterations) assumes that you only `ORDER BY` the offset. That's, for example, why handling errors works the way it does, above: when a new error comes in, it _can_ only have an offset that is larger than the error that we re-hydrated from persist, so it always takes precedence. Ordering by a custom column breaks that.

To fix that, we'd have to graft the `ORDER BY` field on to errors, which feel brittle enough as they are. :see_no_evil: 

Galaxy-brain idea, where I'm not at all sure we should do it. We could have a new implementation of the UPSERT machinery that is only used for new sources when a custom `ORDER BY` is used. We could fix our issues around errors and other things.

Author: To clarify further, if say an Error is persisted, we sort of lose the kafka metadata, since those don't get persisted. But for the incoming kafka stream, we will actually have the metadata, so we will have some order value. So, it's more like a comparison between `None` (when hydrating state from persist) with `Some(ts)` (value coming from kafka) and the latter taking precedence.

Author: > We could have a new implementation of the UPSERT machinery that is only used for new sources when a custom ORDER BY is used.

Does it mean that the upsert operator will only have the kafka ingestion as the input and there will be no read from persist? 

Actually limiting it like that **would** simplify things 😬.

# Rollout
This feature will be rolled out behind the unsafe flag, then upgraded to only being available behind a LaunchDarkly flag.

## Testing and Observability

The custom ordering can be tested via testdrive tests in the CI.

## Lifecycle

> What will it take to promote this to the next stage i.e. Alpha?

Reviewer 3: One thing to watch out for is that as soon as we roll this out to any customers, the syntax will be semi-stable, in that we'll have to be willing to write a migration if we want to change it, and also warn all the customers who were using the old syntax that the syntax has changed.

Reviewer 4: This is only true after its not `unsafe`, right?

- We should work with some users and see how useful this is to them. This would also give us a better idea on how to expand this feature later on.
- Some questions to give us a better idea
    - How is their data structured and what columns will they want to use for the custom order by?
    - What kind of compaction are they using?

# Drawbacks

Users will need to be careful of what kind of compaction they have for the kafka topic. The upsert will do the correct thing based on the data it sees.

> If old data has been compacted away either via time based or key based compaction, and we start ingesting the source after that, the result of upsert might not be accurate.

Reviewer 3: ```suggestion
If old data has been compacted away either via time based or key based compaction, and we start ingesting the source after that, the result of upsert will not include the data Kafka has compacted away, which might violate user expectations.```

If users actually are using this custom order by, it’s likely they are **not** using key based compaction with Kafka because semantically that would be upsert-ing on offset.

# Conclusion and alternatives

An alternative could be waiting for `TRANSFORM USING` which has a much broader scope and could cover this particular scenario. In the meantime, having a smaller scoped custom order by could be useful for interested users.

# Unresolved questions

> - In case the order by value of two entries are same, should it chose one of them deterministically? For eg, say we already have (key1, value1, ts=100) and have incoming (key1, value2, ts=100). Should we override `value1` with `value2` or not?

Reviewer 3: I haven't thought about this too hard, but almost certainly something becomes fundamentally incorrect about the system if we do something nondeterministic here. But it's totally fine to do something _arbitrary_, as long as it's deterministic.

Author: Makes sense. I am leaning towards not updating existing value unless the order by is actually larger (partly because the WIP implementation already works that way). So, in this case `value1` will not be overridden. 

I will actually update the example above with this.

Reviewer 5: I think we have to do something deterministic here. The order in which `(key1, value1)` and `(key1, value2)` arrive will be non-deterministic when users create the same source twice. And I think two distinct sources with the same definition should yield the same results.

Maybe we can fall back to offset as a tie-breaker?

Reviewer 4: Falling back to offset seems like the only reasonable choice here!

Author: Removed this as unresolved question. Added an example above to show the tie-break using offset.
